x/ibc/02-client/handler.go

```diff
@@ -7,32 +7,49 @@ import (
 	evidenceexported "github.com/cosmos/cosmos-sdk/x/evidence/exported"
 	"github.com/cosmos/cosmos-sdk/x/ibc/02-client/exported"
 	"github.com/cosmos/cosmos-sdk/x/ibc/02-client/types"
+	ibctmtypes "github.com/cosmos/cosmos-sdk/x/ibc/07-tendermint/types"
 )
 
 // HandleMsgCreateClient defines the sdk.Handler for MsgCreateClient
-func HandleMsgCreateClient(ctx sdk.Context, k Keeper, msg MsgCreateClient) (*sdk.Result, error) {
-	clientType := exported.ClientTypeFromString(msg.ClientType)
-	if clientType == 0 {
-		return nil, sdkerrors.Wrap(ErrInvalidClientType, msg.ClientType)
+func HandleMsgCreateClient(ctx sdk.Context, k Keeper, msg exported.MsgCreateClient) (*sdk.Result, error) {
+	clientType := exported.ClientTypeFromString(msg.GetClientType())
+	switch clientType {
+	case 0:
+		return nil, sdkerrors.Wrap(ErrInvalidClientType, msg.GetClientType())
+	case 1:
+		tmMsg, ok := msg.(ibctmtypes.MsgCreateClient)
+		if !ok {
+			return nil, sdkerrors.Wrap(ErrInvalidClientType, "Msg is not a Tendermint CreateClient msg")
+		}
+		clientState, err := ibctmtypes.InitializeFromMsg(tmMsg)
+		if err != nil {
+			return nil, err
+		}
+		_, err = k.CreateClient(
+			ctx, clientState, msg.GetConsensusState(),
+		)
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, sdkerrors.Wrap(ErrInvalidClientType, msg.GetClientType())
 	}
 
-	_, err := k.CreateClient(
-		ctx, msg.ClientID, clientType, msg.ConsensusState, msg.TrustingPeriod, msg.UnbondingPeriod,
-	)
-	if err != nil {
-		return nil, err
+	attributes := make([]sdk.Attribute, len(msg.GetSigners())+1)
+	attributes[0] = sdk.NewAttribute(sdk.AttributeKeyModule, AttributeValueCategory)
+	for i, signer := range msg.GetSigners() {
```

Comment:
Would there ever be more than one signer? All this message should require is a fee payment.

Reply:
This function used to take a concrete msg type that had a single signer
Now it takes in an interface that extends `sdk.Msg`

The sdk.Msg interface can have multiple signers so I added the ability to have multiple signer attributes.
In the concrete implementation in ICS-7 there is still a single signer. I don't enforce this behavior in the msg interface because it seems unnecessary

Reply:
Gotcha, seems fine.